floris/parallel_floris_model_2.py

```diff
@@ -0,0 +1,309 @@
+            if self.return_turbine_powers_only:
+                with self._PoolExecutor(self.max_workers) as p:
+                    self._turbine_powers_split = p.starmap(
+                        _parallel_run_powers_only,
+                        parallel_run_inputs
+                    )
+            else:
+                with self._PoolExecutor(self.max_workers) as p:
+                    self._fmodels_split = p.starmap(_parallel_run, parallel_run_inputs)
+            t2 = timerpc()
+            self._postprocessing()
+            self.core.farm.finalize(self.core.grid.unsorted_indices)
+            self.core.state = State.USED
+            t3 = timerpc()
+        if self.print_timings:
+            print("===============================================================================")
+            if self.interface is None:
+                print(f"Total time spent for serial calculation (interface=None): {t1 - t0:.3f} s")
+            else:
+                print(
+                    "Total time spent for parallel calculation "
+                    f"({self.max_workers} workers): {t3-t0:.3f} s"
+                )
+                print(f"  Time spent in parallel preprocessing: {t1-t0:.3f} s")
+                print(f"  Time spent in parallel loop execution: {t2-t1:.3f} s.")
+                print(f"  Time spent in parallel postprocessing: {t3-t2:.3f} s")
+
+    def _preprocessing(self):
+        # Split over the wind conditions
+        n_wind_condition_splits = self.n_wind_condition_splits
+        n_wind_condition_splits = np.min(
+            [n_wind_condition_splits, self.core.flow_field.n_findex]
+        )
+
+        # Prepare the input arguments for parallel execution
+        fmodel_dict = self.core.as_dict()
+        wind_condition_id_splits = np.array_split(
+            np.arange(self.core.flow_field.n_findex),
+            n_wind_condition_splits,
+        )
+        multiargs = []
+        for wc_id_split in wind_condition_id_splits:
+            # for ws_id_split in wind_speed_id_splits:
+            fmodel_dict_split = copy.deepcopy(fmodel_dict)
+            wind_directions = self.core.flow_field.wind_directions[wc_id_split]
+            wind_speeds = self.core.flow_field.wind_speeds[wc_id_split]
+            turbulence_intensities = self.core.flow_field.turbulence_intensities[wc_id_split]
+
+            # Extract and format all control setpoints as a dict that can be unpacked later
+            control_setpoints_subset = {
+                "yaw_angles": self.core.farm.yaw_angles[wc_id_split, :],
+                "power_setpoints": self.core.farm.power_setpoints[wc_id_split, :],
+                "awc_modes": self.core.farm.awc_modes[wc_id_split, :],
+                "awc_amplitudes": self.core.farm.awc_amplitudes[wc_id_split, :],
+                "awc_frequencies": self.core.farm.awc_frequencies[wc_id_split, :],
+            }
+
+            # Do I also need to hangle splitting the wind data object?
```

Comment:
typo

Reply:
Also, we can make some tests to ensure works with WindRose etc.,

Reply:
Agreed. I'm hopeful that it will, since the wind data objects are only actually used in preprocssing (in `set()`) and postprocessing (in `get_...`), but I'll add some tests to check.

Reply:
I've now added those tests and removed this comment.